// @ux/button/examples/index.stories.js
import React, { useState } from 'react';
import Button from '../src';
import '../src/styles';

import Star from '@ux/icon/star';
import '@ux/icon/plus/index.css';
import '@ux/icon/chevron-up/index.css';
import '@ux/icon/star/index.css';


const Collection = (args) => {
  const [active, setActive] = useState(args.active);
  const title = `${args.design.charAt(0).toUpperCase() + args.design.slice(1)} ${args.text}`;

  return (<>
    <Button { ...args } text={ title } />

    <Button { ...args } text={ `${title} Disabled` } disabled={ true }/>

    <Button { ...args } text={ `${title} as select` } as='select' className='test-with-select'/>

    <Button { ...args } text={ `${title} as cta` } as='cta' className='test-with-cta'/>

    <Button { ...args } text={ `${title} as external` } as='external' className='test-with-external'/>

    <Button text={ `${title} as toggle` } as='toggle' active={ active } onClick={ () => setActive(!active) } />

    <Button { ...args } text={ `${title} as link` } href='#'/>

    <Button
      { ...args }
      icon={ <Star/> }
      className='test-with-icon-only'
      // eslint-disable-next-line no-undefined
      text={ undefined }
      aria-label='Favorite this'/>

    <Button { ...args } text={ `${title} with icon` } icon={ <Star/> } className='test-with-icon'/>

    <Button
      { ...args }
      text={ null }
      icon={ <Star/> }
      as='select'
      className='test-with-select-icon'
      title='Select star type'
      aria-label='Select star type'/>

    <Button { ...args } stretch={ true } text={ `${title} stretched` } icon={ <Star/> }/>
  </>);
};

export default {
  title: 'Components/Button',
  component: Button,
  args: {
    design: 'primary',
    size: 'default',
    text: 'Button'
  },
  argTypes: {
    size: { control: 'select', options: ['default', 'small'] },
    design: { control: 'select', options: Button.designs }
  },
  render: Collection
};

export const Primary = {};

export const Small = {
  args: {
    size: 'small'
  }
};

export const Inline = {
  args: {
    design: 'inline'
  }
};

export const Secondary = {
  args: {
    design: 'secondary'
  }
};

export const Stateless = {
  args: {
    design: 'stateless'
  }
};

export const Critical = {
  args: {
    design: 'critical'
  }
};
