// @ux/dialog/src/styles.js
import { css } from 'linaria';
import Intents from '@ux/intents';
import '@ux/button/styles';
import '@ux/text/styles';
import '@ux/sibling-set/styles';
import '@ux/space/styles';
import '@ux/icon/x/styles';
import '@ux/icon/checkmark/styles';
import '@ux/icon/information/styles';
import '@ux/icon/comment/styles';
import '@ux/icon/alert/styles';
import '@ux/icon/help/styles';
import '@ux/icon/star/styles';

const intents = new Intents();

export default css`
  :global() {
    .ux-dialog-shell {
      display: inline-flex;
      gap: calc(${intents.ux.box.density} * 2);
      box-sizing: border-box;
      position: relative;

      &.ux-dialog--stretch {
        display: flex;
        min-width: 100%;
      }
    }

    .ux-dialog-icon {
      display: inline-flex;
      flex-shrink: 0;
      align-self: baseline;
    }

    .ux-dialog-details {
      flex: 1;
    }

    .ux-dialog-float {
      float: right;
      width: 2rem;
      height: 2rem;
      > * {
        position: absolute;
        top: 0;
        right: 0;
        color: currentColor;
      }
    }

    .ux-dialog-title {
      overflow-x: auto;
      overflow-y: clip;
      margin-block-end: calc(${intents.ux.box.density} * 2);
      > * {
        margin: 0;
      }
    }

    .ux-dialog-content {
      overflow: visible;
    }

    .ux-dialog-actions {
      margin-block-start: calc(${intents.ux.box.density} * 2);
    }

    .ux-dialog--contrast .ux-dialog-actions {
      filter: grayscale(1);
      -webkit-font-smoothing: antialiased;
      -moz-osx-font-smoothing: grayscale;
    }
  }
`;
