// @ux/dialog/examples/stacking.stories.js
import Dialog from '../src';
import text from '@ux/text';
import Button from '@ux/button';
import React from 'react';

export { default as default } from './index.stories.js';

export function Stacking() {

  const title = <text.h2 text='Red pill or blue pill? This decision is totally up to you; choose wisely' size={ 2 } />;
  const actions = (
    <>
      <Button text='Red pill' design='critical'/>
      <Button text='Blue pill' design='primary'/>
    </>
  );
  // eslint-disable-next-line no-console
  const onClose = () => console.log('close!');
  const content = (
    <ul style={{ margin: 0, padding: 0, listStyle: 'none' }}>
      <li>You take the blue pill—the story ends,
      you wake up in your bed and believe whatever you want to believe.</li>
      <li>You take the red pill—you stay in Wonderland,
      and I show you how deep the rabbit hole goes.</li>
    </ul>
  );

  const flyoutStyle = {
    position: 'absolute',
    height: 'calc(100% - 1rem)',
    width: 'calc(100% - 1rem)',
    background: 'purple',
    padding: '1rem',
    color: 'yellow',
    margin: '.5rem'
  };

  return (
    <div style={{ display: 'flex', flexWrap: 'wrap', gap: '3rem', position: 'relative' }}>

      <Dialog
        emphasis='critical'
        title={ title }
        actions={ actions }
        onClose={ onClose }>
        { content }
      </Dialog>

      <div style={ flyoutStyle }>
        This should appear on top of the dialog because it is the last element in the DOM.
        If the dialog is given a z-index, it will appear above this.
      </div>
    </div>
  );
}
